File: Kickstart 2018/Practice Round/C.py
def solve():
    N = int(input())
    trip = {}
    srcs = set()
    dsts = set()
    for _ in range(N):
        src = input()
        dst = input()
        trip[src] = dst
        srcs.add(src)
        dsts.add(dst)
    start = [i for i in srcs if i not in dsts][0]
    result = []
    while start in trip:
        result.append('{}-{}'.format(start, trip[start]))
        start = trip[start]
    print(' '.join(result))

if __name__ == '__main__':
    T = int(input())
    for t in range(T):
        print('Case #{}: '.format(t+1), end='')
        solve()

File: Kickstart 2018/Practice Round/C.cpp
#include "../../../stdc++.h"
#define _ ios_base::sync_with_stdio(0);cin.tie(0);
#define repeat(i, n) for (int i = 0; (i) < (n); ++(i))
using namespace std;

void solve() {
    int N; cin >> N;
    map<string, string> trip;
    set<string> srcs;
    set<string> dsts;
    repeat(i, N) {
        string src, dst; cin >> src >> dst;
        trip.insert(pair<string, string> (src, dst));
        srcs.insert(src);
        dsts.insert(dst);
    }
    string start;
    for(auto src : srcs) {
        if(!dsts.count(src)) {
            start = src;
            break;
        }
    }
    while(trip.count(start)) {
        cout << " " << start << "-" << trip[start];
        start = trip[start];
    }
    cout << endl;
}

int main(void) { _
    int T; cin >> T;
    repeat(t, T) {
        cout << "Case #" << t+1 << ":";
        solve();
    }
    return 0;
}